backends/p4tools/cmake/Z3.cmake

```diff
@@ -0,0 +1,85 @@
+macro(p4tools_obtain_z3)
+  option(TOOLS_USE_PREINSTALLED_Z3 "Look for a preinstalled version of Z3 instead of installing a prebuilt binary using FetchContent." OFF)
+
+  if(TOOLS_USE_PREINSTALLED_Z3)
+    # We need a fairly recent version of Z3.
+    set(Z3_MIN_VERSION "4.8.14")
+    # But 4.12+ is currently broken with libGC
+    set(Z3_MAX_VERSION_EXCL "4.12")
+    find_package(Z3 ${Z3_MIN_VERSION} REQUIRED)
+
+    if(NOT DEFINED Z3_VERSION_STRING OR ${Z3_VERSION_STRING} VERSION_LESS ${Z3_MIN_VERSION})
+      message(FATAL_ERROR "The minimum required Z3 version is ${Z3_MIN_VERSION}. Has ${Z3_VERSION_STRING}.")
+    endif()
+    if(${Z3_VERSION_STRING} VERSION_GREATER_EQUAL ${Z3_MAX_VERSION_EXCL})
+      message(FATAL_ERROR "The Z3 version has to be lower than ${Z3_MAX_VERSION_EXCL} (the latter currently does no work with libGC). Has ${Z3_VERSION_STRING}.")
+    endif()
+    # Set variables for later consumption.
+    set(P4TOOLS_Z3_LIB z3::z3)
+    set(P4TOOLS_Z3_INCLUDE_DIR ${Z3_INCLUDE_DIR})
+  else()
+    # Pull in a specific version of Z3 and link against it.
+    set(P4TOOLS_Z3_VERSION "4.13.0")
```

Comment:
It seems odd that Z#_MAX_VERSION_EXCL above is 4.12, but you are explicitly pulling in 4.13.0 here.  Do you expect one of those two things to change before this PR is ready to merge?

Reply:
Z3 versions above 4.12 added [malloc_usable_size](https://github.com/Z3Prover/z3/blob/b2b3bab246840f077b21f4b0c5a6e667acb8bdc4/src/util/memory_manager.cpp#L278) which causes problems with P4C's garbage collector. I have not figured out a fix for BDWGC yet. 

When we build from scratch we do not have this problem because we can patch the Z3 installation.

Reply:
That should be a comment clarifying that.

Reply:
There is some more work to be done first, need to figure out whether it is possible to patch bdwgc instead.